aws/rust-runtime/aws-types/src/profile/parse.rs

```diff
@@ -0,0 +1,333 @@
+
+    /// Error message
+    message: String,
+}
+
+impl Display for ProfileParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "error parsing {} on line {}:\n  {}",
+            self.location.path, self.location.line_number, self.message
+        )
+    }
+}
+
+/// Validate that a line represents a valid subproperty
+///
+/// - Subproperties looks like regular properties (`k=v`) that are nested within an existing property.
+/// - Subproperties must be validated for compatibility with other SDKs, but they are not actually
+/// parsed into structured data.
+fn validate_subproperty(value: &str, location: Location) -> Result<(), ProfileParseError> {
+    if value.trim_matches(WHITESPACE).is_empty() {
+        Ok(())
+    } else {
+        parse_property_line(value)
+            .map_err(|err| err.into_error("sub-property", location))
+            .map(|_| ())
+    }
+}
+
+fn is_empty_line(line: &str) -> bool {
+    line.trim_matches(WHITESPACE).is_empty()
+}
+
+fn is_comment_line(line: &str) -> bool {
+    line.starts_with(COMMENT)
+}
+
+/// Parser for profile files
+struct Parser<'a> {
+    /// In-progress profile representation
+    data: RawProfileSet<'a>,
+
+    /// Parser state
+    state: State<'a>,
+
+    /// Parser source location
+    ///
+    /// Location is tracked to facilitate error reporting
+    location: Location,
+}
+
+enum State<'a> {
+    Starting,
+    ReadingProfile {
+        profile: &'a str,
+        property: Option<&'a str>,
+        is_subproperty: bool,
+    },
+}
+
+/// Parse `file` into a `RawProfileSet`
+pub fn parse_profile_file(file: &File) -> Result<RawProfileSet, ProfileParseError> {
+    let mut parser = Parser {
+        data: HashMap::new(),
+        state: State::Starting,
+        location: Location {
+            line_number: 0,
+            path: (&file.path).to_string(),
+        },
+    };
+    parser.parse_profile(&file.contents)?;
+    Ok(parser.data)
+}
+
+impl<'a> Parser<'a> {
+    /// Parse `file` containing profile data into `self.data`.
+    fn parse_profile(&mut self, file: &'a str) -> Result<(), ProfileParseError> {
+        for (line_number, line) in file.lines().enumerate() {
+            self.location.line_number = line_number + 1; // store a 1-indexed line number
+            if is_empty_line(line) || is_comment_line(line) {
+                continue;
+            }
+            if line.starts_with('[') {
+                self.read_profile_line(line)?;
+            } else if line.starts_with(WHITESPACE) {
+                self.read_property_continuation(line)?;
+            } else {
+                self.read_property_line(line)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Parse a property line like `a = b`
+    ///
+    /// A property line is only valid when we're within a profile definition, `[profile foo]`
+    fn read_property_line(&mut self, line: &'a str) -> Result<(), ProfileParseError> {
```

Comment:
Should `read_property_line` and `read_property_continuation` both be using `prepare_line`?

Edit: I see that it's eventually called in `parse_property_line`. Would it be clearer if the top-level `parse_profile` prepared the line before anything else?

Reply:
heh, it can't do that because lines that start with whitespace are significant. I will take a pass at reorganizing this though